strawberryfields/program.py

```diff
@@ -0,0 +1,897 @@
+  It is a mapping from subsystem indices to lists of Commands touching that subsystem,
+  where each list is temporally ordered.
+* Finally, the quantum circuit can be represented using a DAG by making each Command a node,
+  and drawing an edge from each Command to all its immediate followers along each wire it touches.
+  It can be converted back into a command queue by popping a maximal element until the graph
+  is empty, that is, consuming it in a topological order.
+  Note that a topological order is not always unique, there may be several equivalent topological orders.
+
+.. currentmodule:: strawberryfields.program.Program
+
+The three representations can be converted to each other
+using the methods :func:`_list_to_grid`, :func:`_grid_to_DAG` and :func:`_DAG_to_list`.
+
+
+Optimizer
+---------
+
+The purpose of the optimizer part of the compiler is to simplify the circuit
+to make it cheaper and faster to execute. Different backends might require
+different types of optimization, but in general the fewer operations a circuit has,
+the faster it should run. The optimizer thus should convert the circuit into a
+simpler circuit while preserving the probability distributions of the measurement results.
+The optimization utilizes the abstract algebraic properties of the gates,
+and in no point should require a matrix representation.
+
+Currently the optimization is very simple. It
+
+* merges neighboring gates belonging to the same gate family and sharing the same set of subsystems
+* cancels neighboring pairs of a gate and its inverse
+
+.. currentmodule:: strawberryfields.program
+
+
+Code details
+~~~~~~~~~~~~
+
+"""
+# pylint: disable=too-many-instance-attributes,attribute-defined-outside-init
+
+# todo: Avoid issues with Program contexts and threading,
+# cf. _pydecimal.py in the python standard distribution.
+
+from collections.abc import Sequence
+import copy
+import functools
+import numbers
+
+import networkx as nx
+
+import strawberryfields.circuitdrawer
+
+
+
+def _print_list(i, q, print_fn=print):
+    "For debugging."
+    # pylint: disable=unreachable
+    return
+    print_fn('i: {},  len: {}   '.format(i, len(q)), end='')
+    for x in q:
+        print_fn(x.op, ', ', end='')
+    print_fn()
+
+def _convert(func):
+    r"""Decorator for converting user defined functions to a :class:`RegRefTransform`.
+
+    This allows classical processing of measured qumodes values.
+
+    Example usage:
+
+    .. code-block:: python
+
+        @convert
+        def F(x):
+            # some classical processing of x
+            return f(x)
+
+        with eng:
+            MeasureX       | q[0]
+            Dgate(F(q[0])) | q[1]
+
+    Args:
+        func (function): function to be converted to a :class:`RegRefTransform`.
+    """
+    @functools.wraps(func)
+    def wrapper(*args):
+        "Unused docstring."
+        return RegRefTransform(args, func)
+    return wrapper
+
+
+# TODO replace this mock backend capability database
```

Comment:
Looks good for now 👍 but we'll definitely need something more complicated to allow us to also specify restricted topologies